client/state/reader/posts/selectors.js

```diff
@@ -11,42 +17,30 @@ import { filter, keyBy } from 'lodash';
  * @param  {String}  postGlobalId Post global ID
  * @return {Object} Post
  */
-export function getPost( state, postGlobalId ) {
+export function getPostById( state, postGlobalId ) {
 	return state.reader.posts.items[ postGlobalId ];
 }
 
-let previousItems = null,
-	postMapBySiteAndPost = {};
-/**
- * Get a single post by site ID and post ID
- * @param  {Object} state  Global state tree
- * @param  {Number} siteId Site ID
- * @param  {Number} postId Post ID
- * @return {object}        Post object. Undefined if missing.
- */
-export function getPostBySiteAndId( state, siteId, postId ) {
-	if ( state.reader.posts.items !== previousItems ) {
-		/*
-		 * Create a memoized map of posts by site ID and post ID
-		 * Invalidate it when the items map changes to a new instance
-		 * Didn't use createSelector here because it memoizes per argument,
-		 * which would create a cache per item, instead of one map with all the
-		 * items. Doing it this way is still fast (map lookup instead of array search)
-		 * and saves memory in the common case.
-		 */
-		const items = state.reader.posts.items,
-			// only internal (wpcom / jetpack) posts have a valid site_ID and post ID
-			internalPosts = filter(
-				items,
-				post => post && post.site_ID && post.ID && ! post.is_external
-			);
-
-		postMapBySiteAndPost = keyBy( internalPosts, post => {
-			return `${ post.site_ID }-${ post.ID }`;
-		} );
+const getPostMapByPostKey = treeSelect(
+	state => [ state.reader.posts.items ],
+	( [ posts ] ) => keyBy( posts, post => keyToString( keyForPost( post ) ) )
+);
 
-		previousItems = items;
+export const getPostByKey = ( state, postKey ) => {
+	if ( ! postKey ) {
+		return null;
 	}
+	const postMap = getPostMapByPostKey( state );
+	return postMap[ keyToString( postKey ) ];
+};
 
-	return postMapBySiteAndPost[ `${ siteId }-${ postId }` ];
-}
+export const getPostsByKeys = treeSelect(
```

Comment:
why a separate treeSelect here? seems like the cache-key is overly specific and won't match much?

Reply:
just thinking about this vs reusing the `getPostMapByPostKey` and looking up via that.

Reply:
ah, because otherwise the returned array would change on every invocation and cause a re-render... hrm.

Reply:
maybe instead of resolving the post key to a post here, we add a new HoC that takes a postKey, resolves the key to a post, and then passes the post down to a wrapped component?

Reply:
then we can pass the array of postKeys, which doesn't change.

Reply:
I'm confused as to what the HoC would do exactly...

accept an array of postKeys or just an individual postKey?
How can we avoid creating a new array of posts at some level?

Reply:
ah. 

the right solution would be to refactor combined cards + recommended posts to not need the array of posts.  They should work solely off postKeys and then the individual subcomponents `CombinedCard` and `RelatedPosts` would be connected and would call `getPostForKey`